src/metabase/driver/postgres.clj

```diff
@@ -298,6 +302,73 @@
   [driver [_ arg]]
   (sql.qp/->honeysql driver [:percentile arg 0.5]))
 
+(defn- datetime-diff-helper [x y unit]
+  (case unit
+    (:year :day)
+    (hx/cast
+     :integer
+     (hsql/call
+      :date_part
+      (hx/literal unit)
+      (hsql/call
+       (case unit :year :age :day :-)
+       (hsql/call :date_trunc (hsql/raw "'day'") y)
+       (hsql/call :date_trunc (hsql/raw "'day'") x))))
+
+    :week
+    (hx// (datetime-diff-helper x y :day) 7)
+
+    :month
+    (hx/cast
+     :integer
+     (hx/+
+      (hx/* 12 (datetime-diff-helper x y :year))
+      (hsql/call
+       :date_part
+       (hsql/raw "'month'")
+       (hsql/call
+        :age
+        (hsql/call :date_trunc (hsql/raw "'day'") y)
+        (hsql/call :date_trunc (hsql/raw "'day'") x)))))
+
+    (:hour :minute :second)
+    (let [ex            (hsql/call :extract :epoch (hx/->timestamptz x))
+          ey            (hsql/call :extract :epoch (hx/->timestamptz y))
+          positive-diff (fn [a b]
+                          (hx/cast
+                           :integer
+                           (hx/floor
+                            (if (= unit :second)
+                              (hx/- b a)
+                              (hx// (hx/- b a) (case unit :hour 3600 :minute 60))))))]
+      (hsql/call :case (hsql/call :<= ex ey) (positive-diff ex ey) :else (hx/* -1 (positive-diff ey ex))))
+
+    (throw (ex-info (tru "Invalid datetime-diff unit: {0}" unit)
+                    {:valid-units [:year :month :week :day :hour :minute :second]
+                     :bad-unit unit}))))
+
+;; x : date/datetime
+;; y : date/datetime
+;; unit : year | month | week | day | hour | minute | second
+;;
+;; The result is positive if x <= y, and negative otherwise.
+;;
+;; Days, weeks, months, and years are only counted if they are whole to the "day".
+;; For example, `datetimeDiff("2022-01-30", "2022-02-28", "month")` returns 0 months.
+;;
+;; If the values are datetimes, the time doesn't matter for these units.
+;; For example, `datetimeDiff("2022-01-01T09:00:00", "2022-01-02T08:00:00", "day")` returns 1 day even though it is less than 24 hours.
+;;
+;; Hours, minutes, and seconds are only counted if they are whole.
+;; For example, datetimeDiff("2022-01-01T01:00:30", "2022-01-01T02:00:29", "hour") returns 0 hours.
+(defmethod sql.qp/->honeysql [:postgres :datetime-diff]
+  [driver [_ x y unit]]
+  (let [x (sql.qp/->honeysql driver (cond-> x
+                                      (string? x) u.date/parse))
```

Comment:
I think it's a sign that something is wrong if we're having the driver code parse timestamp string literals inside the clauses themselves. This should already be done at this point -- QP middleware should have taken care of it. It's better to handle this stuff once in QP middleware than to have to write the same code 20 times for 20 different drivers to handle it.

I think what you probably want to do is add some code in the `wrap-value-literals/wrap-value-literals-in-mbql` to handle `:datetime-diff` and write a test just for that middleware to make sure it's doing the right thing